match_p.h

```diff
@@ -246,11 +250,37 @@ static inline unsigned longest_match(deflate_state *const s, IPos cur_match) {
         match++;
 
         do {
-        } while (*(uint16_t *)(scan += 2)== *(uint16_t *)(match += 2) &&
-                 *(uint16_t *)(scan += 2)== *(uint16_t *)(match += 2) &&
-                 *(uint16_t *)(scan += 2)== *(uint16_t *)(match += 2) &&
-                 *(uint16_t *)(scan += 2)== *(uint16_t *)(match += 2) &&
-                 scan < strend);
+            uint16_t mval, sval;
+
+            MEMCPY(&mval, match, sizeof(mval));
+            MEMCPY(&sval, scan, sizeof(sval));
+            if (mval != sval)
+              break;
+            match += sizeof(mval);
+            scan += sizeof(sval);
+
+            MEMCPY(&mval, match, sizeof(mval));
+            MEMCPY(&sval, scan, sizeof(sval));
+            if (mval != sval)
+              break;
+            match += sizeof(mval);
+            scan += sizeof(sval);
+
+            MEMCPY(&mval, match, sizeof(mval));
+            MEMCPY(&sval, scan, sizeof(sval));
+            if (mval != sval)
+              break;
+            match += sizeof(mval);
+            scan += sizeof(sval);
+
+            MEMCPY(&mval, match, sizeof(mval));
+            MEMCPY(&sval, scan, sizeof(sval));
+            if (mval != sval)
+              break;
+            match += sizeof(mval);
+            scan += sizeof(sval);
+
+        } while (scan < strend);
```

Comment:
This introduces unnecessary code duplication... It should compare the memory locations directly using largest possible data type or use only one pair of MEMCPY per iteration...

Reply:
I will fix this in a follow-up patch that will change the performance.
We will need to special case big/little endian when iterating through the bytes of the loaded `uint64_t`.

Reply:
Using oprimized memcmp() you wouldn't need to byte swap anything... Basically compiler will automatically use temporary variable correctly,